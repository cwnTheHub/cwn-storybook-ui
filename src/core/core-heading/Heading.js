import React, { forwardRef } from "react";
import PropTypes from "prop-types";

import styled from "styled-components";
import {
  helveticaNeueLight45,
  helveticaNeueThin35,
  helveticaNeueMedium65,
  wordBreak,
  baseSupSubScripts,
} from "../../shared-typography/typography";
import media from "../core-responsive/media";
import { colorWhite, colorText, colorSecondary } from "../core-colours/colours";
import { safeRest } from "../../util-helpers";

const HeadingLevels = {
  h1: {
    ...helveticaNeueLight45,
    fontSize: "1.75rem",
    lineHeight: "1.29", // 36px
    letterSpacing: "-1.6px",
    ...media.from("md").css({
      ...helveticaNeueThin35,
      fontSize: "2.75rem",
      lineHeight: "1.18",
      letterSpacing: "0",
    }),
    sup: {
      ...baseSupSubScripts,
      fontSize: "1.25rem",
      top: "-1em",
      ...media.from("md").css({ fontSize: "1.25rem", top: "-1.3em" }),
    },
  },
  h2: {
    ...helveticaNeueLight45,
    fontSize: "1.5rem",
    lineHeight: "1.33", // 30px
    letterSpacing: "-0.7px",
    ...media.from("md").css({
      fontSize: "1.75rem",
      lineHeight: "1.29",
      letterSpacing: "-0.8px",
    }),
    sup: {
      ...baseSupSubScripts,
      fontSize: "1rem",
      top: "-0.8em",
      ...media.from("md").css({ fontSize: "1rem", top: "-0.7em" }),
    },
  },
  h3: {
    ...helveticaNeueMedium65,
    fontSize: "1.25rem",
    lineHeight: "1.4", // 28px
    letterSpacing: "-0.6px",
    sup: {
      ...baseSupSubScripts,
      fontSize: "0.875rem",
      top: "-0.5em",
    },
  },
  h4: {
    ...helveticaNeueMedium65,
    fontSize: "1rem",
    lineHeight: "1.25", // 20px
    letterSpacing: "-0.6px",
    sup: {
      ...baseSupSubScripts,
      fontSize: "0.875rem",
      top: "-0.5em",
    },
  },
};
export const StyledHeading = styled.h1(wordBreak, ({ level, invert }) => {
  const baseColor =
    level === "h1" || level === "h2" ? colorSecondary : colorText;
  const color = invert ? colorWhite : baseColor;
  return {
    color,
    ...HeadingLevels[`${level}`],
    "& > span": {
      letterSpacing: "inherit",
    },
  };
});

const Heading = forwardRef(
  ({ level, tag = level, invert, children, ...rest }, ref) => {
    return (
      <StyledHeading
        {...safeRest(rest)}
        ref={ref}
        as={tag}
        level={level}
        invert={invert}
        data-testid="heading"
      >
        {children}
      </StyledHeading>
    );
  }
);

Heading.displayName = "Heading";

Heading.propTypes = {
  level: PropTypes.oneOf(["h1", "h2", "h3", "h4"]).isRequired,
  tag: PropTypes.oneOf(["h1", "h2", "h3", "h4", "div", "span"]),
  invert: PropTypes.bool,
  children: PropTypes.node.isRequired,
};
Heading.defaultProps = {
  invert: false,
  tag: undefined,
};

export default Heading;
