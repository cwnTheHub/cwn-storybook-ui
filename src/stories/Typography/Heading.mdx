import {
  Canvas,
  Meta,
  Title,
  Subtitle,
  Description,
  Primary,
  Controls,
  Stories,
} from "@storybook/blocks";

import * as HeadingStories from "./Heading.stories";

<Meta of={HeadingStories} />
<Title />

`import {Heading} from '@nds_core/nds'`

### PROPS & METHODS
<Controls/>

<Canvas of={HeadingStories.Minimal} />

#### Usage criteria

- Be mindful of SEO and accessibility
- Organize headers in a nested order: h1, then h2, then h3
- Try **not** to skip levels where possible: h1, then h3
- Use one h1 per page
- The semantic HTML tag and visual appearance of the Heading component can be controlled using the `level` and `tag` props together
- For Hero overlays, Headlines, or Promo blocks, use [`DisplayHeading`](?path=/docs/core-components-typography-displayheading--documentation)
- For subtext in the Hero overlays, or Headlines, use [`Text`](?path=/docs/core-components-typography-text--documentation)

#### Separating HTML tag from visual style

The `Heading` component allows developers, designers, and content editors to represent headings with a certain style
while rendering a HTML tag of their choice. It's important to maintain a correct order of heading level and not skip
levels; for example, have an `<h2>` HTML element directly follows an `<h1>`, or an `<h3>` directly follows and `<h2>`. This allows users of
assistive technology to navigate through content predictably since headings represent information
hierarchy.

In the example below, the content would appear at the top of a page, as a Promo.

- The first line of text (Consider this...) appears as a visual heading 4, but renders with an `<h1>` tag
- The second line of text (Getting membership...) appears as a visual heading 1, but renders with an `<h2>`

When we provide different heading tags, the heading hierarchy is in order for assistive technology, but we are able
to have a different visual hierarchy of text.

<Canvas of ={HeadingStories.Example}/>
