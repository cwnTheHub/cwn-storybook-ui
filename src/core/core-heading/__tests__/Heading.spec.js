import React from "react";
import { mount, render } from "enzyme";

import Heading from "../Heading";
import {
  colorWhite,
  colorSecondary,
  colorText,
} from "../../core-colours/colours";

describe("Heading", () => {
  const defaultProps = {
    level: "h1",
  };

  const doMount = (overrides = {}) => {
    const heading = mount(
      <Heading {...defaultProps} {...overrides}>
        The content
      </Heading>
    );
    return {
      heading: heading.find(heading.props().tag || heading.props().level),
      styledComponent: heading.find('[data-testid="heading"]'),
    };
  };

  it("renders", () => {
    const heading = render(<Heading level="h2">The heading</Heading>);

    expect(heading).toMatchSnapshot();
  });

  it("renders text", () => {
    const { heading } = doMount();

    expect(heading.text()).toBe("The content");
  });

  it("renders a heading in the specified level", () => {
    const { heading } = doMount({ level: "h3" });
    expect(heading.name()).toBe("h3");
  });

  it("renders a heading in the specified level and specified tag, when level and tag are different", () => {
    const { heading } = doMount({ level: "h4", tag: "h3" });
    expect(heading.name()).toEqual("h3");
  });

  it("renders as a span", () => {
    const { heading } = doMount({ level: "h4", tag: "span" });
    expect(heading.name()).toEqual("span");
  });

  it("renders as a div", () => {
    const { heading } = doMount({ level: "h4", tag: "div" });
    expect(heading.name()).toBe("div");
  });

  it("renders a heading in the same tag as level if tag is not specified", () => {
    const { heading } = doMount({ level: "h4" });

    expect(heading.name()).toEqual("h4");
  });

  describe("colour", () => {
    it("can be inverted", () => {
      const { styledComponent } = doMount({ invert: true });

      expect(styledComponent.get(0).props.invert).toBe(true);
    });

    it("is secondary for h1 and h2", () => {
      let heading = doMount({ level: "h1" });
      expect(heading.styledComponent.get(0).props.level).toBe("h1");

      heading = doMount({ level: "h2" });
      expect(heading.styledComponent.get(0).props.level).toBe("h2");
    });

    it("is default for h3 and h4", () => {
      let heading = doMount({ level: "h3" });
      expect(heading.styledComponent.get(0).props.level).toBe("h3");

      heading = doMount({ level: "h4" });
      expect(heading.styledComponent.get(0).props.level).toBe("h4");
    });
  });

  it("passes additional attributes to heading element", () => {
    const { heading } = doMount({ id: "the-heading", tabIndex: 1 });

    expect(heading.props().id).toEqual("the-heading");
    expect(heading.props()["tabIndex"]).toEqual(1);
  });

  it("does not allow custom CSS", () => {
    const { heading } = doMount({
      className: "my-custom-class",
      style: { color: "hotpink" },
    });

    expect(heading.props()["className"]).not.toBe("my-custom-class");
    expect(heading.props()["style"]).toBe(undefined);
  });
});
