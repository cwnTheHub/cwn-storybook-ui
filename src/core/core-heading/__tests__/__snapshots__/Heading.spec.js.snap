// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading renders 1`] = `
<h2
  class="sc-aXZVg hFgKLd"
  data-testid="heading"
>
  The heading
</h2>
`;
